crates/p2p/src/swarm/setup/upgrade.rs

```diff
@@ -0,0 +1,124 @@
+//! Protocol upgrade implementations for the setup phase.
+//!
+//! This module provides the upgrade implementations for both inbound and outbound
+//! setup protocols, handling the serialization and exchange of setup messages
+//! over libp2p streams.
+
+use std::{future::Future, pin::Pin};
+
+use asynchronous_codec::{Framed, JsonCodec};
+use futures::{SinkExt, StreamExt};
+use libp2p::{
+    InboundUpgrade, OutboundUpgrade, PeerId, Stream, core::UpgradeInfo, identity::PublicKey,
+};
+
+use crate::{
+    signer::ApplicationSigner,
+    swarm::{errors::SetupUpgradeError, message::SignedMessage},
+};
+
+/// Inbound upgrade for handling incoming setup requests.
+///
+/// This upgrade processes incoming setup messages from remote peers.
+#[derive(Debug, Clone)]
+pub struct InboundSetupUpgrade;
+
+impl InboundSetupUpgrade {
+    /// Creates a new inbound setup upgrade.
+    pub(crate) const fn new() -> Self {
+        Self
+    }
+}
+
+impl UpgradeInfo for InboundSetupUpgrade {
+    type Info = &'static str;
+    type InfoIter = std::iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        std::iter::once("/handshake/1.0.0")
+    }
+}
+
+impl InboundUpgrade<Stream> for InboundSetupUpgrade {
+    type Output = SignedMessage;
+    type Error = SetupUpgradeError;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn upgrade_inbound(self, stream: Stream, _: Self::Info) -> Self::Future {
+        Box::pin(async move {
+            let mut framed = Framed::new(stream, JsonCodec::<SignedMessage, SignedMessage>::new());
+
+            match StreamExt::next(&mut framed).await {
+                Some(Ok(signed_message)) => Ok(signed_message),
+                Some(Err(e)) => Err(SetupUpgradeError::JsonCodec(e.into())),
+                None => Err(SetupUpgradeError::UnexpectedStreamClose),
+            }
+        })
+    }
+}
+
+/// Outbound upgrade for initiating handshake requests.
+///
+/// This upgrade sends the setup messages to remote peers.
+#[derive(Clone, Debug)]
+pub struct OutboundSetupUpgrade<S: ApplicationSigner> {
+    pub(crate) app_public_key: PublicKey,
+    pub(crate) local_peer_id: PeerId,
+    pub(crate) remote_peer_id: PeerId,
```

Comment:
docstrings?
What is the difference between `local_peer_id` and `remote_peer_id`?
The docstrings should tell this.

Reply:
Renamed to `local_transport_id` and `remote_transport_id`. Documented once in message.rs.